File: java/src/main/java/com/fengtian/qianer/atcoder/abc/abc131/ABC131_C.java
package com.fengtian.qianer.atcoder.abc.abc131;

import java.util.Scanner;

public class ABC131_C {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        final long A = sc.nextLong();
        final long B = sc.nextLong();
        final long C = sc.nextLong();
        final long D = sc.nextLong();
        final long L = lcm(C, D);
        System.out.println((B - B / C - B / D + B / L) - ((A-1) - (A-1) / C - (A-1) / D + (A-1) / L));
    }

    private static long lcm(long a, long b) {
        return a * b / gcd(a, b);
    }

    private static long gcd(long a, long b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }
}

File: kotlin/src/abc/abc131/ABC131_C.kt
package abc.abc131

fun main(args: Array<String>) {
    val (a, b, c, d) = readLine()!!.split(" ").map(String::toLong)

    println((b-(a-1)) - (b/c-(a-1)/c) - (b/d-(a-1)/d) + b/lcm(c, d)-(a-1)/lcm(c, d))
}

fun lcm(a: Long, b: Long): Long {
    return a * b / gcd(a, b)
}

fun gcd(a: Long, b: Long): Long {
    return if (b == 0L) a else gcd(b, a % b)
}